pallets/gear/src/lib.rs

```diff
@@ -1379,112 +1483,79 @@ pub mod pallet {
-                Error::<T>::ValueLessThanMinimal
-            );
-
-            let schedule = T::Schedule::get();
-
-            ensure!(
-                code.len() as u32 <= schedule.limits.code_len,
-                Error::<T>::CodeTooLarge
-            );
-
-            let code = Code::try_new(code, schedule.instruction_weights.version, |module| {
-                schedule.rules(module)
-            })
-            .map_err(|e| {
-                log::debug!("Code failed to load: {:?}", e);
-                Error::<T>::FailedToConstructProgram
-            })?;
-
-            ensure!(
-                code.code().len() as u32 <= schedule.limits.code_len,
-                Error::<T>::CodeTooLarge
-            );
-
-            let code_and_id = CodeAndId::new(code);
-
-            let packet = InitPacket::new_with_gas(
+            let code_and_id = Self::check_code(code)?;
+            let packet = Self::init_packet(
+                who.clone(),
                 code_and_id.code_id(),
                 salt,
                 init_payload,
                 gas_limit,
-                value.unique_saturated_into(),
-            );
-
-            let program_id = packet.destination();
-            // Make sure there is no program with such id in program storage
-            ensure!(
-                !GearProgramPallet::<T>::program_exists(program_id),
-                Error::<T>::ProgramAlreadyExists
-            );
-
-            let reserve_fee = T::GasPrice::gas_price(gas_limit);
-
-            // First we reserve enough funds on the account to pay for `gas_limit`
-            // and to transfer declared value.
-            CurrencyOf::<T>::reserve(&who, reserve_fee + value)
-                .map_err(|_| Error::<T>::NotEnoughBalanceForReserve)?;
-
-            let origin = who.clone().into_origin();
-
-            let code_id = code_and_id.code_id();
+                value,
+            )?;
 
-            // By that call we follow the guarantee that we have in `Self::submit_code` -
-            // if there's code in storage, there's also metadata for it.
-            if let Ok(code_hash) = Self::set_code_with_metadata(code_and_id, origin) {
-                // TODO: replace this temporary (`None`) value
-                // for expiration block number with properly
-                // calculated one (issues #646 and #969).
-                Self::deposit_event(Event::CodeChanged {
-                    id: code_hash,
-                    change: CodeChangeKind::Active { expiration: None },
-                });
+            if !T::CodeStorage::exists(code_and_id.code_id()) {
+                // By that call we follow the guarantee that we have in `Self::upload_code` -
+                // if there's code in storage, there's also metadata for it.
+                if let Ok(code_hash) =
+                    Self::set_code_with_metadata(code_and_id, who.clone().into_origin())
+                {
```

Comment:
if code doesn't exist, the failure here must fail the extrinsic

Reply:
Discussed with @breathx. Need to propagate error here. So please remove ```if let Ok(code_hash)``` and introduce instead ```let code_hash = <call>()?;```.